### PR TITLE
Fixes the usability issues with the comment section delete and edit

### LINKED ISSUE #7281: Some issues with cancel button and delete icon when comments are edited
The _Cancel_ button shown when a comment is edited overflows in the next comment, as can be seen in the screenshot below:
comment-edit-cancel-overflow

It is more evident when the next comment is also in edition mode, as in that case the _Cancel_ button of the previous comment and the _Delete_ icon of the next comment overlap:
comment-edit-delete

Regarding the _Delete_ icon, it is shown only when the cursor is on the comment element. As the _Cancel_ button is a child of the comment element but is visually outside it that causes the _Delete_ icon to appear when hovering on the _Cancel_ button and disappear when moving the cursor to the sides of that button, which is a bit strange.

Besides that, the _Delete_ icon is aligned to the top and right border of the comment element, and it has no padding nor margin. When the cursor enters the comment element on the top right the _Delete_ icon appears immediately under the cursor, which also feels strange; it would be better if it had some margin so the icon appeared slightly before the cursor was on top of it.

Also, I have not tested it, but I guess that showing the _Delete_ icon only when the cursor is on the comment could be problematic on touch screens. Similarly, the area to interact with the icon should be increased on touch screens, as 16x16 is not very usable in that case.

@nextcloud/designers

### PR DESCRIPTION
**Pull request by Abijeet**
Greetings,

I've made the following changes in the comments section,

1. Increased the size of the delete button to take up 44x44px.
2. Now showing the delete button at all times when editing to avoid usability issues with touch screens.
3. Edit icon is also shown only at all times, rather than only on hover, which fixes usability issues.
4. Edit icon has been made 44x44px.
5. Submit icon has been made 44x44px.
6. Fixed div being added placed directly under `ul` during editing of comments.
7. Saving a non-modified comment now reverts the edit window.
8. Reduced the gap between comments a *wee* bit.
9. Fixed issues with focus event now working properly for edit and delete.

Please let me know if any modifications need to be made.

### Comment section
screenshot-2018-1-12 files - nextcloud 1

### Editing a comment
screenshot-2018-1-12 files - nextcloud

Closes #7281

Signed-off-by: Abijeet <abijeetpatro@gmail.com>

/cc @nextcloud/designers

### CODE REVIEW

**Comment by Abijeet** [comment]
Also, while checking the code I could not find why we needed this - `_onClickComment` - 

https://github.com/nextcloud/server/blob/b248c53c6a8a2d420256921ca4889fc553ce8fee/apps/comments/js/commentstabview.js#L526-L532

**Comment by danxuliu** [comment]
>  why we needed this - _onClickComment -

[Long messages](https://github.com/nextcloud/server/blob/b248c53c6a8a2d420256921ca4889fc553ce8fee/apps/comments/js/commentstabview.js#L714-L716) are initially [created with the `collapsed` CSS class](https://github.com/nextcloud/server/blob/b248c53c6a8a2d420256921ca4889fc553ce8fee/apps/comments/js/commentstabview.js#L43), which [sets a maximum size for the comment and fades its contents](https://github.com/nextcloud/server/blob/b248c53c6a8a2d420256921ca4889fc553ce8fee/apps/comments/css/comments.css#L80-L113); clicking on the comment removes the `collapsed` class and fully shows it (or at least it should, I have not checked it ;-) ).

**Comment by Abijeet** [comment]
@danxuliu - Thanks, it does work as intended.

**Comment by rullzer** [comment]
Summoning reviewers :)

**Comment by blizzz** [comment]
Would be happy to have feedback from the design team

**Comment by pixelipo** [comment]
it would be much easier to review if the source branch would be in nextcloud/server

File: apps/comments/css/comments.css [hunk 1]
```diff
@@ -18,19 +18,18 @@
 }
 
 #commentsTabView .newCommentForm .message {
-	width: calc(100% - 81px); /* 36 (left margin) + 30 (right padding) + 15 (right padding of surrounding box) */
+	width: calc(100% - 88px);
```
Comment (pixelipo) [inline apps/comments/css/comments.css]: @Abijeet can you please leave a comment regarding where this number comes from?

File: apps/comments/css/comments.css [hunk 2]
```diff
@@ -18,19 +18,18 @@
...
 	margin-left: 36px;
```
Comment (pixelipo) [inline apps/comments/css/comments.css]: IMO, it's the parent div (`newCommentForm`) that should be moved 36px to the left

File: apps/comments/css/comments.css [hunk 3]
```diff
@@ -18,19 +18,18 @@
...
-	padding-right: 30px;
 	display: block;
```
Comment (pixelipo) [inline apps/comments/css/comments.css]: should be `inline-block`

File: apps/comments/css/comments.css (hunk 2 above)
Comment (pixelipo) [reply apps/comments/css/comments.css]: If you make this margin 1 or 2 px smaller, then the `.submit` icon can fit without any absolute positioning

File: apps/comments/css/comments.css [hunk 4]
```diff
@@ -18,19 +18,18 @@
...
 }
 
 #commentsTabView .newCommentForm .submit {
 	position: absolute;
```
Comment (pixelipo) [inline apps/comments/css/comments.css]: lines 27-30 unnecessary (see my comments above)

File: apps/comments/css/comments.css [hunk 5]
```diff
@@ -73,11 +72,18 @@
 
 #commentsTabView .comment {
 	position: relative;
-	margin-bottom: 30px;
+	margin-bottom: 20px;
```
Comment (pixelipo) [inline apps/comments/css/comments.css]: suggestion: add something like `border-top: 1px solid $color-border;` to have a better separation. In fact, if you add that, it becomes obvious that this line should really be `padding: 10px 0;`.

I also noticed that the `.comment` class is applied twice - once to the `div.newCommentRow` element and then again to it's `li.comment` children. This doesn't sound right....
image

**Review by pixelipo — CHANGES_REQUESTED** [review]
1. tooltip is now too high:
image
2. Delete icon seems to jump around: before hover:
image
after hover:
image
3. I find this to be a big pile of mess:
image
 - why are these 3 actions placed in a column instead of inline, like everywhere else in NC?
 - if the should stay vertical, why are they not vertically aligned?
 - why is cancel not an icon, too? "x" is the classic solution used across NC apps

**Comment by danxuliu** [comment]
@pixelipo 
>  - why are these 3 actions placed in a column instead of inline, like everywhere else in NC?
>  - if the should stay vertical, why are they not vertically aligned?
>  - why is cancel not an icon, too? "x" is the classic solution used across NC apps

I guess that the answer to those three questions is "_because it was already that way before this pull request_" ;-)

Having said that, yes, I too think that those buttons could be improved. Maybe something like [what @jancborchardt said](https://github.com/nextcloud/server/issues/7281#issuecomment-356235124) would be good enough:
>  Maybe it makes more sense to put the cancel button above, in the same line as the avatar / delete button for uniformity.

More specifically, I would replace the _Edit_ button by the _Cancel_ button with the classic "x". I think that makes sense from a _behavioural_ point of view (same place for enter edition and exit edition), and this way it would be far enough from the delete button too to press it by mistake (although the delete button would still be near the confirm button...).

File: apps/comments/css/comments.css (hunk 5 above)
Comment (danxuliu) [reply apps/comments/css/comments.css]: >  I also noticed that the .comment class is applied twice - once to the div.newCommentRow element and then again to it's li.comment children. This doesn't sound right....

In your screenshot the `li` is not a child of the `div`; the `li` is a child of the `ul`, which is a sibling of the `div`. If I am not mistaken the `.comment` class is fine that way, as it is applied to each comment in the list, and also to the "_New comment_" area.

**Comment by Abijeet** [comment]
@pixelipo - Thanks for the review on this one, I should have been a little more careful and thorough with my changes. I was unsure of making too many modifications.

> it would be much easier to review if the source branch would be in nextcloud/server

Will do this from next time onwards. I'll also take care of the comments you've left on the code. 

Regarding re-adjusting the layout, what do you guys think of the following that I made after reviewing comments left by @jancborchardt and @danxuliu ,

files nextcloud

**Comment by danxuliu** [comment]
@Abijeet Thanks a lot for taking care of all this! :-)

Just a little note (for the future ;-) ) regarding the commits. When possible please try to use different commits for changes not related between them (for example, increasing the touch area of the icons and exiting the edition mode when saving a comment that was not modified). This makes easier to review the commits before merging them and, maybe more important, in some time in the future when someone needs to understand the code and how it evolved ;-) But do not worry too much about that; it is just one of my little nitpicks, most people do not seem to care about those things :-P

>  Regarding re-adjusting the layout, what do you guys think of the following that I made

I would not change the send button in order to keep a consistent UI between sending a new comment and editing an existing one.

I was thinking on something like this instead:
comment-edit-delete-icon

However I have no special preference for that approach; it is just an additional idea.

By the way, regarding
>  Delete icon seems to jump around

I have noticed that in Firefox, although it seems to be working fine in Chromium. I do not know if there is something wrong with the CSS or if it is a browser bug.

**Comment by Abijeet** [comment]
Hi @danxuliu and @pixelipo, I've been a little more thorough with my changes this time and have taken some liberties. Here's a full set of issues that have been fixed,

1. Increased the size of the delete, edit, close and submit buttons to take up 44x44px.
2. Now showing the delete button at all times when editing to avoid usability issues with touch screens.
3. Edit icon is also shown at all times, rather than only on hover.
4. Reduced the gap between comments a wee bit.
5. Fixed issues with focus event now working properly for edit and delete.
6. Removed absolute positioning of elements for alignment.
7. Fixed issue with tooltips becoming too high due to padding around actions. Occurred due to my changes.
8. Changed the position of the submit and close icons.
9. Fixed issue with jumping of the delete icon in Firefox. Occurred due to my changes.
10. Fixed issue with wrapping of content due to long author names. Occurred due to my changes.
11. Fixed issue with longer comments not appearing properly. This _might_ have occurred due to my changes.
12. Converts the .css file to .scss file, and adds border-top to comments.
13. Fixes div being directly under ul during editing of comments.
14. Fixes issue where saving a non-modified comment kept the edit window open.

## Tests

- I've tested to ensure that the loading symbols appear properly. 
- I've tested to ensure that comments look fine with longer author names.


## Response to comments

> Just a little note (for the future ;-) ) regarding the commits. When possible please try to use different commits for changes not related between them (for example, increasing the touch area of the icons and exiting the edition mode when saving a comment that was not modified). 

I've undone my previous commits and tried to do this as much as I could for this pull request, but I'll be more careful from next time on wards.

>  - why are these 3 actions placed in a column instead of inline, like everywhere else in NC?
>  - if the should stay vertical, why are they not vertically aligned?
>  - why is cancel not an icon, too? "x" is the classic solution used across NC apps

This has been changed. Please check the images below.

> @Abijeet can you please leave a comment regarding where this number comes from?

Added comments

> IMO, it's the parent div (newCommentForm) that should be moved 36px to the left

Done.

> If you make this margin 1 or 2 px smaller, then the .submit icon can fit without any absolute positioning
> should be inline-block
> lines 27-30 unnecessary (see my comments above)

Removed and implemented without absolute positioning.

> suggestion: add something like border-top: 1px solid $color-border; to have a better separation. In fact, if you add that, it becomes obvious that this line should really be padding: 10px 0;.

Changed to use padding. Added border-top but not for the new comment. Please check screenshot.

Please let me know if there are any other changes, or anything else is not up-to the mark.

## Screenshots

### Normal mode
screenshot-2018-1-19 files - nextcloud 1

### Edit mode
screenshot-2018-1-19 files - nextcloud 2

### Long usernames
screenshot-2018-1-19 files - nextcloud 3

**Comment by jancborchardt** [comment]
There are 3 icons without label, and especially the x and delete next to each other are confusing. Hence I'd put the Cancel and Delete into a 3-dot-menu, with labels.

Otherwise great! :)

**Comment by Abijeet** [comment]
@jancborchardt - I agree with @pixelipo, I don't think the cancel button should go under a dropdown. Not sure where to put the delete button though, next to the edit? 

Would what @danxuliu suggested make more sense? 

> comment-edit-delete-icon

**Comment by juliushaertl** [comment]
I see no real need for the delete button while editing the text. How about adding a 3-dots-menu for the default view and add edit/delete to it. On the editing view we could just show the submit/cancel buttons then.

**Comment by Abijeet** [comment]
I like the approach suggested by @juliushaertl, I'll go ahead with that.

**Comment by Abijeet** [comment]
Hey all,

I've made the changes. I've added a new View that is the dropdown for the edit and delete options under a menu. 

I've slight reservations about the code [here](https://github.com/nextcloud/server/pull/7800/files#diff-07287428c5872cf7258d96b38fb1237eR408). I figured keeping it simple and the modification to the minimum was the right way to go about it. 

Please let me know if you'd like me to change that or anything else.

cc @nextcloud/designers 

## Screenshots

### Dropdown visible
comment-open-menu
----

### Dropdown in case of large username
comment-menu-open-right
----

### Editing a comment
comment-edit
----

**Comment by Abijeet** [comment]
@nextcloud/designers - Any thoughts?

**Comment by skjnldsv** [comment]
Awesome!!
I just think the full username is a bit too much! Could we set a max with and use an ellipsis? :)

**Comment by Abijeet** [comment]
@skjnldsv - Done. Would you like a tooltip too? We'd have to probably show the tooltip only if it crosses a certain number of characters.

screenshot-localhost-2018 02 25-20-09-53

**Comment by jancborchardt** [comment]
Really great stuff! Only 2 things:

- The 3-dot popover should be centered below the 3-dot icon. Currently it's too far to the right.
- Make sure the submit as well as the cancel icon have titles "Submit comment"/"Cancel" which are shown by JS so they are immediately visible?


(@juliushaertl good suggestion btw. :)

**Comment by Abijeet** [comment]
@jancborchardt - 

> The 3-dot popover should be centered below the 3-dot icon. Currently it's too far to the right.

Its not centered because on smaller resolutions having a centered popover is causing the popover to go out of the window when the username is slightly long. Hence, based on the length of the username I'm deciding whether to show the popover to the right or the left.

> Make sure the submit as well as the cancel icon have titles "Submit comment"/"Cancel" which are shown by JS so they are immediately visible?

Good catch, cancel button has, submit does not, I'll add that.

**Comment by Abijeet** [comment]
> Make sure the submit as well as the cancel icon have titles "Submit comment"/"Cancel" which are shown by JS so they are immediately visible?

This is done. Left the same tool-tips as were before - 
- *Post* when adding a new comment.
- *Save* when updating a comment.

@nextcloud/designers - Please review and merge at your convenience.

**Review by skjnldsv — COMMENTED** [review]
Could we have a loader on comment delete? :)

Otherwise: ok from me! :+1:

**Comment by Abijeet** [comment]
@skjnldsv - Loader has been added already. Please see the screenshot below and the respective code [here](https://github.com/nextcloud/server/pull/7800/files#diff-07287428c5872cf7258d96b38fb1237eR582). I just tweaked the code to hide the menu before triggering the click action. 

Cheers.

loader-delete

**Comment by skjnldsv** [comment]
> @skjnldsv - Loader has been added already. Please see the screenshot below and the respective code here. I just tweaked the code to hide the menu before triggering the click action.

My bad, with the opacity AND the loader, I couldn't see anything :/
Could you makes sure the loader is not impacted by the opacity? :) Maybe apply the opacity to .message and not the full comment? :)

**Comment by jancborchardt** [comment]
Other than @skjnldsv's comment this seems great now! :) Outstanding work @Abijeet, happy to have you on the Nextcloud Design team! :)

**Comment by skjnldsv** [comment]
@Abijeet Keep us updated :)

**Comment by Abijeet** [comment]
@skjnldsv - This is done. Now when a comment is disabled, only the comment "message" opacity is reduced.

**Review by skjnldsv — APPROVED** [review]

**Comment by Abijeet** [comment]
@nextcloud/designers - Requesting a review.

**Comment by enoch85** [comment]
Great improvement! Thanks @Abijeet :)

**Comment by skjnldsv** [comment]
JSunit test are failing @Abijeet :)
Still a bit work to do ;)

https://drone.nextcloud.com/nextcloud/server/5701/214

**Comment by Abijeet** [comment]
Thank you @skjnldsv , I should have checked rather than assuming their failure was unrelated. Unfortunately it is the end of Sunday already. I will review and fix these this week.

**Comment by Abijeet** [comment]
I think I've fixed the issues with the comments related test cases failing. Going to close and reopen the issue to force the test cases to run again.

**Comment by Abijeet** [comment]
Hmph, the reopening did not retrigger the CI. @skjnldsv - Any ideas on how to proceed from here? It seems that the docker connection failed.

**Comment by skjnldsv** [comment]
@Abijeet you need to rebase to master and force push your work :)

**Comment by Abijeet** [comment]
@skjnldsv - Done, but that doesn't seem to have helped.

**Comment by skjnldsv** [comment]
@Abijeet yes it did, but our drone have a little issue! :)
We'll restart it today ;)

**Comment by Abijeet** [comment]
@skjnldsv - How do I get the CI to trigger again?

**Comment by skjnldsv** [comment]
> @skjnldsv - How do I get the CI to trigger again?

Login on drone, right menu, restart.
I just restarted it! :)

**Comment by MorrisJobke** [comment]
> @skjnldsv - How do I get the CI to trigger again?
> Login on drone, right menu, restart.
> I just restarted it! :) https://drone.nextcloud.com/nextcloud/server/6234

Failures are unrelated.

File: apps/comments/css/comments.scss [hunk 6]
```diff
@@ -13,53 +13,40 @@
...
 }
 
 #commentsTabView .newCommentForm .message {
-	width: calc(100% - 81px); /* 36 (left margin) + 30 (right padding) + 15 (right padding of surrounding box) */
-	margin-left: 36px;
-	padding-right: 30px;
-	display: block;
+	/* width = 100% - (width of submit button (44px) + margin (3px) + inline-block gap) */
+	width: calc(100% - 52px);
+	display: inline-block;
```
Comment (MorrisJobke) [inline apps/comments/css/comments.scss]: This causes the empty input to have a too big cursor in Safari:

bildschirmfoto 2018-04-03 um 17 47 53

See also this thread https://stackoverflow.com/a/9132362

I could also not figure out how to enforce this without the `display: block;`, `font-size`, `line-height`and `height` all didn't worked out :/

Comment (MorrisJobke) [reply apps/comments/css/comments.scss]: Once the first character is typed all is fine.

**Review by MorrisJobke — APPROVED** [review]
Beside my little problem it looks good 👍 Would be still nice to get this resolved as well before we merge this.

File: apps/comments/css/comments.scss (hunk 6 above)
Comment (MorrisJobke) [reply apps/comments/css/comments.scss]: Merged this and we should fix this little issue in #9150

Comment (Abijeet) [reply apps/comments/css/comments.scss]: Got a new laptop and was setting it up. :)  This kind of took a back seat.

**Pull request merged**